File: elixir/yacht/lib/yacht.ex
defmodule Yacht do
  @type category ::
          :ones
          | :twos
          | :threes
          | :fours
          | :fives
          | :sixes
          | :full_house
          | :four_of_a_kind
          | :little_straight
          | :big_straight
          | :choice
          | :yacht

  @doc """
  Calculate the score of 5 dice using the given category's scoring method.
  """
  @spec score(category :: category(), dice :: [integer]) :: integer

  def score(:ones, dice), do: dice |> Stream.filter(&(&1 == 1)) |> Enum.sum()

  def score(:twos, dice), do: dice |> Stream.filter(&(&1 == 2)) |> Enum.sum()

  def score(:threes, dice), do: dice |> Stream.filter(&(&1 == 3)) |> Enum.sum()

  def score(:fours, dice), do: dice |> Stream.filter(&(&1 == 4)) |> Enum.sum()

  def score(:fives, dice), do: dice |> Stream.filter(&(&1 == 5)) |> Enum.sum()

  def score(:sixes, dice), do: dice |> Stream.filter(&(&1 == 6)) |> Enum.sum()

  def score(:full_house, dice) do
    case Enum.sort(dice) do
      [n, n, n, m, m] when n != m -> Enum.sum(dice)
      [m, m, n, n, n] when n != m -> Enum.sum(dice)
      _ -> 0
    end
  end

  def score(:four_of_a_kind, dice) do
    case Enum.sort(dice) do
      [n, n, n, n, _] -> n * 4
      [_, n, n, n, n] -> n * 4
      _ -> 0
    end
  end

  def score(:little_straight, dice) do
    case Enum.frequencies(dice) do
      %{1 => 1, 2 => 1, 3 => 1, 4 => 1, 5 => 1} -> 30
      _ -> 0
    end
  end

  def score(:big_straight, dice) do
    case Enum.frequencies(dice) do
      %{2 => 1, 3 => 1, 4 => 1, 5 => 1, 6 => 1} -> 30
      _ -> 0
    end
  end

  def score(:choice, dice), do: Enum.sum(dice)

  def score(:yacht, dice) do
    case dice do
      [n, n, n, n, n] -> 50
      _ -> 0
    end
  end

end

File: c/yacht/yacht.c
#include "yacht.h"

int score(const dice_t dice, const category_t category) {
  int dices[6] = {0};
  for (int i = 0; i < 5; ++i) {
    int d = dice.faces[i];
    if (d < 1 || d > 6) {
      return 0;
    }
    ++dices[d - 1];
  }
  int counts[6] = {0};
  for (int i = 0; i < 6; ++i) {
    int count = dices[i];
    ++counts[count];
  }
  int score = 0;
  switch (category) {
  case ONES:
    return dices[0];
  case TWOS:
    return 2 * dices[1];
  case THREES:
    return 3 * dices[2];
  case FOURS:
    return 4 * dices[3];
  case FIVES:
    return 5 * dices[4];
  case SIXES:
    return 6 * dices[5];
  case FULL_HOUSE:
    if (counts[2] != 1 || counts[3] != 1) {
      return 0;
    }
    for (int i = 0; i < 6; ++i) {
      score += (i + 1) * dices[i];
    }
    return score;
  case FOUR_OF_A_KIND:
    if (counts[4] == 0 && counts[5] == 0) {
      return 0;
    }
    for (int i = 0; i < 6; ++i) {
      if (dices[i] >= 4) {
        return (i + 1) * 4;
      }
    }
    return 0;
  case LITTLE_STRAIGHT:
    for (int i = 0; i < 5; ++i) {
      if (dices[i] == 0) {
        return 0;
      }
    }
    return 30;
  case BIG_STRAIGHT:
    for (int i = 1; i < 6; ++i) {
      if (dices[i] == 0) {
        return 0;
      }
    }
    return 30;
  case CHOICE:
    for (int i = 0; i < 6; ++i) {
      score += dices[i] * (i + 1);
    }
    return score;
  case YACHT:
    for (int i = 0; i < 6; ++i) {
      if (dices[i] == 5) {
        return 50;
      }
    }
    return 0;
  }
  return 0;
}
